### ISSUE TITLE
tfjs - trouble running speech commands in React Native CLI app

### ISSUE THREAD

**Issue by sseppola** [tensorflow/tfjs#2785]
To get help from the community, we encourage using Stack Overflow and the [`tensorflow.js`](https://stackoverflow.com/questions/tagged/tensorflow.js) tag.

#### TensorFlow.js version
"@tensorflow-models/speech-commands": "0.4.2",
"@tensorflow/tfjs": "1.5.2",
"@tensorflow/tfjs-react-native": "0.2.3",

#### Browser version
"react-native": "0.61.5",

#### Describe the problem or feature request
I'm having a lot of trouble loading `@tensorflow-models/speech-commands` as it seems to be missing dependencies. I'm fairly certain I initialized the project correctly according to the [setup guide](https://github.com/tensorflow/tfjs/tree/master/tfjs-react-native#setting-up-a-react-native-app-with-tfjs-react-native)

#### Code to reproduce the bug / link to feature request

I initialized the project with React Native CLI
```
$ npx react-native init [projectName] --template react-native-template-typescript
```
and followed https://github.com/tensorflow/tfjs/tree/master/tfjs-react-native#setting-up-a-react-native-app-with-tfjs-react-native

After the setup the app runs fine on both Android and iOS, however, as soon as I try to load `@tensorflow-models/speech-commands` I get problems:

**First:** `error: bundling failed: Error: Unable to resolve module `util` from `node_modules/@tensorflow-models/speech-commands/dist/browser_fft_utils.js`: util could not be found within the project.`
Fixed with `yarn add util`

**Second:** `error: bundling failed: Error: Unable to resolve module `fs` from `node_modules/@tensorflow-models/speech-commands/dist/browser_fft_utils.js`: fs could not be found within the project.`
This one I have not been able to resolve, and it seems to be triggered by the line: `recognizer = speechCommands.create('BROWSER_FFT', 'directional4w');`

The closest issue I found to this was from #1682 in an expo-cli app, and I tried the tip from https://github.com/tensorflow/tfjs/issues/1682#issuecomment-534231135 where I `require('@tensorflow-models/speech-commands')` after `await tf.ready();` without luck.

I'm a bit confused as the naming of the files implies that it runs in the browser, however both `util` and `fs` are node.js modules. I'm expecting there's some polyfilling needed to get this running properly.

Here's the relevant code:
```tsx
import React, {useEffect} from 'react';
import {View, Text} from 'react-native';
import {Audio} from 'expo-av';

import * as tf from '@tensorflow/tfjs';
import '@tensorflow/tfjs-react-native';
import * as speechCommands from '@tensorflow-models/speech-commands';

export default function SpeechCommands() {
  useEffect(() => {
    let recognizer: speechCommands.SpeechCommandRecognizer | undefined;

    const start = async () => {
      await tf.ready();

      // the following line throws an error:
      recognizer = speechCommands.create('BROWSER_FFT', 'directional4w');
    };

    start();

    return () => {
      if (recognizer) {
        recognizer.stopListening();
      }
    };
  }, []);

  return (
    <View>
      <Text>SpeechCommands component</Text>
    </View>
  );
}
```

**Comment by tafsiri** [comment]
You are correct on the polyfilling. This was written before react-native support and I believe uses conditional requires in some places for node modules. Metro however tries to load them all. The quickest resolution may be to try something like https://github.com/tradle/rn-nodeify to shim the various node require statements.

However there is a larger issue that will likely prevent you from using this model. The model is wholly dependent on using the browser based FFT implementation to convert audio into something the model can process. Since this is not present in React Native (AFAIK), you would need some other way to convert the audio into the same representation (and I know of no straightforward way to do this—though it wouldn't be impossible). I would thus pretty much consider this model browser only.

**Comment by sseppola** [comment]
Good tip regarding rn-nodeify!

The larger issue you mention seems tricky as I'm just getting into Tensorflow, could I bounce some ideas off you to decide how to proceed?

As far as I can tell I've got 3 options, ordered from what I think would be easiest:

**Option A**: Use native models
Given that `@tensorflow-models/speech-commands` is for the browser, would it be possible to do the opposite and load a native ios/android model into `@tensorflow/react-native` + `@tensorflow/tfjs`?
I'm thinking I could just use the model from the tensorflow light example: https://github.com/tensorflow/examples/tree/master/lite/examples/speech_commands

**Option B**: Use the native Tensorflow libraries with native models
Start from https://github.com/tensorflow/examples/tree/master/lite/examples/speech_commands and continue native/brownfield app development working with the native Tensorflow library and models directly.

**Option C**: Convert the audio into the same representation
The two other options seem easier for me, so we can get back to what this involves if necessary.

**Comment by tafsiri** [comment]
Option A
We don't support converting from tflite models directly. But the example you pointed to has python training scripts that will output a model that you can try to [convert](https://github.com/tensorflow/tfjs/tree/master/tfjs-converter). This involves setting up tensorflow and getting familiar with the python side (which may not be a problem). However it will still require you to come up with a way to capture audio in react native and convert it to the representation expected by that model.

Option B
As I understand it, this means using [tf-lite](https://www.tensorflow.org/lite) directly in an Android/iOS app developed with the native stacks for those operating systems? If so, this will be the most straightforward assuming you are comfortable with native app development.

Option C
Its like option above, the hard part is getting the audio in react-native and converting it to a model appropriate format.

Other than that I don't know your overall goals, but I would definitely suggest looking at [tflite](https://www.tensorflow.org/lite) and seeing if it matches your requirements. 

Option A or C would be great for other JavaScript/React Native devs as you could publish a model on npm more tailored to RN! But I can't really estimate the time or complexity involved.

**Comment by sseppola** [comment]
Aha, now I'm starting to see some of my misconceptions.

I thought I could reuse the same model between tfjs and tflite, and that @tensorflow/react-native bridged into native, now that I read the [description](https://github.com/tensorflow/tfjs/tree/master/tfjs-react-native) again I understand it executes in js!

I'm working on a PoC ios/android app that needs:
1. to recognize a handful of words (< 10, here the speech_commands is enough)
2. learn a few custom words (common for all users, not per-user)
3. adapt to the user's voice/pronunciation
4. run on-device for extended sessions in the background

As far as I understand now this is possible with both `tflite` and `tfjs`, and point 2 and 3 will require "transfer learning" from the existing model.
Can you confirm this? And do you know which of tflite and tfjs would be more battery intensive?

I don't know native development, but I'm probably going to do a PoC in Swift and tflite to get familiarity with the tools, and to find highly applicable guides, tutorials, and examples. I'll then reassess how I can bring this into react-native when I have a better understanding, at this point it seems a bit too complicated to convert the model and figure out the audio capture.

**Comment by sseppola** [comment]
@tafsiri you've been more than helpful, thank you so much :)

Feel free to close the issue, I'm happy to figure out the rest myself

**Comment by tafsiri** [comment]
Sounds good. To your point 3, transfer learning is what you would want for specific user adaptation. Point 2 may or may not use transfer learning. However you might be able to get pretty far even without transfer learning. I can't really comment on the battery side. All the best with your project!

**Comment by tyagit** [comment]
Hey @sseppola did you find a way to reach out to a conclusion for your use case?
```
I'm working on a PoC ios/android app that needs:

1. to recognize a handful of words (< 10, here the speech_commands is enough)
2. learn a few custom words (common for all users, not per-user)
3. adapt to the user's voice/pronunciation
4. run on-device for extended sessions in the background
```
If yes, could you please share the process on how to achieve this?

**Comment by sseppola** [comment]
Hi @tyagit 

Given the criteria above I gave up on using tfjs in favor of tflite. These are the points I considered:
- I had an example to base my code on
- tflite has been around longer, so it should be easier to learn/have better support
- tflite is more capable than tfjs, and so I considered it more likely to support the personalization features by the time I need it

I copied the files from the tflite speech commands example into my RN project, and manually integrated across the RN bridge it to handle start, stop, and emit inference events. It was quite the learning experience.

Here's some thoughts on how this achieved the criteria:

2. As far as learning custom words I'm collecting more samples and training a general model periodically. At this point I do not have acceptable accuracy for custom words but I expect to achieve it with time.

3. I have not done personalization, nor had an urgent need for it yet. Having submitted my own voice samples and trained the general model it seemed it got a little more accurate. I'll probably get back to this at some point, but the performance is good enough as it is.

4. My app runs fine in the background for extended periods. I expect tflite to be more efficient/faster as it runs native code vs crossing the bridge N times per second to run inference in JS. That said, however, I have not measured it.

**Comment by EricMarcantonioIR** [comment]
Hey @sseppola !

Did you find a way to transform audio streams into something that browser_fft can understand? I have some ideas, but before I go and deep dive, I was wondering if you have a solution.

Thank you!
Eric

**Comment by sseppola** [comment]
I did not, ended up using using tflite and writing native integrations for what I needed.

**Comment by jamesbalcombe83** [comment]
@sseppola and/or @EricMarcantonioIR my team and I are trying to achieve something very similar for a final project in a fullstack bootcamp. We are using react-native. I wonder if you would be willing to have a discussion about your approach and some of the dependencies you used to achieve speech recognition. It is proving very challenging for us so far, and there is very little resource to support our goal. We would greatly appreciate your help.

**Comment by EricMarcantonioIR** [comment]
@jamesbalcombe83 Hey, I explained my thoughts in the above comment. I wanted to use RN, but there are no good solutions. If you just have a couple commands, Teachable Machine is a good idea. Tensorflow also has packages that let you use tflite models. Here is the link that I used. https://teachablemachine.withgoogle.com/train/audio
 
Here is the other link I looked at: https://github.com/tensorflow/examples/tree/master/lite/examples/sound_classification

If you want RN, you are going to have to write the packages yourself. I am debating creating an opensource project for it

Cheers
